catalyst_voices/packages/catalyst_voices_services/lib/src/crypto/vault_crypto_service.dart

```diff
@@ -0,0 +1,233 @@
+        throw const CryptoDataMalformed();
+      }
+
+      // Extract the version, algorithm ID
+      final version = data[0];
+      final algorithmId = data[1];
+
+      if (version != _currentVersion) {
+        throw CryptoVersionUnsupported('Version $version');
+      }
+
+      if (algorithmId != _currentAlgorithmId) {
+        throw CryptoAlgorithmUnsupported('Algorithm $version');
+      }
+
+      final algorithm = AesGcm.with256bits(nonceLength: _viLength);
+      final secretKey = SecretKey(key.sublist(_saltLength));
+
+      final encryptedData = data.sublist(2);
+
+      final secretBox = SecretBox.fromConcatenation(
+        encryptedData,
+        nonceLength: _viLength,
+        macLength: algorithm.macAlgorithm.macLength,
+      );
+
+      final decryptedData = await algorithm
+          .decrypt(secretBox, secretKey: secretKey)
+          .then(Uint8List.fromList)
+          .onError<SecretBoxAuthenticationError>(
+            (_, __) => throw const CryptoAuthenticationException(),
+          );
+
+      // Verify checksum/marker
+      final checksum = _checksum;
+      if (decryptedData.length < checksum.length) {
+        throw const CryptoDataMalformed('Data is too short');
+      }
+      final originalDataLength = decryptedData.length - checksum.length;
+      final originalData = decryptedData.sublist(0, originalDataLength);
+      final extractedChecksum = decryptedData.sublist(originalDataLength);
+
+      if (!listEquals(checksum, extractedChecksum)) {
+        throw const CryptoDataMalformed('Checksum mismatch');
+      }
+
+      return originalData;
+    }
+
+    if (kDebugMode) {
+      return _benchmark(run(), debugLabel: 'Decrypt');
+    } else {
+      return run();
+    }
+  }
+
+  @override
+  Future<Uint8List> encrypt(
+    Uint8List data, {
+    required Uint8List key,
+  }) {
+    Future<Uint8List> run() async {
+      final algorithm = AesGcm.with256bits(nonceLength: _viLength);
+      final secretKey = SecretKey(key.sublist(_saltLength));
+
+      final checksum = _checksum;
+      final combinedData = Uint8List.fromList([...data, ...checksum]);
+
+      final secretBox = await algorithm.encrypt(
+        combinedData,
+        secretKey: secretKey,
+        nonce: null,
+        aad: [],
+        possibleBuffer: null,
+      );
+
+      final concatenation = secretBox.concatenation();
+
+      final metadata =
+          Uint8List.fromList([_currentVersion, _currentAlgorithmId]);
+
+      final result = Uint8List.fromList([
+        ...metadata,
+        ...concatenation,
+      ]);
+
+      return result;
+    }
+
+    if (kDebugMode) {
+      return _benchmark(run(), debugLabel: 'Encrypt');
+    } else {
+      return run();
+    }
+  }
+
+  /// Builds list with [length] and random bytes in it.
+  Uint8List _generateRandomList({
+    required int length,
+  }) {
+    final list = Uint8List(length);
+
+    for (var i = 0; i < length; i++) {
+      list[i] = _random.nextInt(16);
```

Comment:
Uint8 has value ranges from 0 till 255. Why 16?

Reply:
No particular reason. Salt usual uses low number but it can be anything. May set it to 255